sql/analyzer/resolve_ctes.go

```diff
@@ -280,104 +283,100 @@ func hoistRecursiveCte(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope,
 	default:
-		return nil, sql.ErrInvalidRecursiveCteRecursiveQuery.New(n)
 	}
 
-	return plan.NewRecursiveCte(union.Left(), union.Right(), sq.Name(), sq.Columns, deduplicate), nil
-}
-
-func resolveRecursiveCte(ctx *sql.Context, a *Analyzer, node sql.Node, sq sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
-	rCte := node.(*plan.RecursiveCte)
-	if rCte == nil {
-		return node, transform.SameTree, nil
+	if hasTable(name, union.Right()) {
+		l, r := splitRecursiveCteUnion(name, union.Left())
+		if r == nil {
+			return union.Left(), union.Right()
+		}
+		return l, plan.NewUnion(r, union.Right(), union.Distinct, union.Limit, union.SortFields)
 	}
 
-	newInit, _, err := a.analyzeThroughBatch(ctx, rCte.Init, scope, "default-rules", sel)
-	if err != nil {
-		return node, transform.SameTree, err
-	}
+	return plan.NewUnion(union.Left(), union.Right(), union.Distinct, union.Limit, union.SortFields), nil
+}
 
-	// create recursive schema from initial projection cols and names
-	var outputProj []sql.Expression
-	switch n := newInit.(type) {
-	case *plan.Project:
-		outputProj = n.Projections
-	case *plan.GroupBy:
-		outputProj = n.SelectedExprs
+// resolveRecursiveCte resolves the static left node of the CTE to perform
+// 1) schema discovery for the CTE and recursive right half, and 2) replace
+// recursive UnresolvedTable references with resolved RecursiveTable nodes.
+func resolveRecursiveCte(
+	ctx *sql.Context,
+	a *Analyzer,
+	rCte *plan.RecursiveCte,
+	scope *Scope,
+	sel RuleSelector,
+) (sql.Node, error) {
+	newInit, _, err := a.analyzeThroughBatch(ctx, rCte.Left(), scope, "default-rules", sel)
+	if err != nil {
+		return rCte, err
 	}
 
-	schema := make(sql.Schema, len(outputProj))
-	var name string
-	for i, p := range outputProj {
-		switch c := p.(type) {
-		case *expression.Alias, *expression.GetField:
-			name = c.(sql.Nameable).Name()
-		case *expression.Literal, sql.Aggregation:
-			name = c.String()
-		default:
-			return nil, transform.SameTree, fmt.Errorf("failed to resolve or unsupported field: %v", p)
-		}
-		if i < len(rCte.Columns) {
-			name = rCte.Columns[i]
-		}
-		schema[i] = &sql.Column{
-			Name:     name,
-			Source:   rCte.Name(),
-			Type:     p.Type(),
-			Nullable: p.IsNullable(),
+	recSch := make(sql.Schema, len(newInit.Schema()))
+	for i, c := range newInit.Schema() {
+		newC := c.Copy()
+		if len(rCte.Columns) > 0 {
+			newC.Name = rCte.Columns[i]
 		}
+		newC.Source = rCte.Name()
+		recSch[i] = newC
 	}
 
-	// resolve recursive table with proper schema
-	rTable := plan.NewRecursiveTable(rCte.Name(), schema)
-
-	// replace recursive table refs
-	newRec, sameR, err := transform.Node(rCte.Rec, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
-		switch t := n.(type) {
+	// replace recursive table refs, cannot do this until we have schema
+	// TODO does ResolvedTable need a schema? should we replace in prior step?
```

Comment:
ResolvedTable by definition has a schema, so I don't know what this comment means

Reply:
deleting comment, was trying to figure out if we could lazily load the recursive schema, but I don't think we can